File: config/base_station/wind_sensor.py
# Copyright 2020 Makani Technologies LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Wind sensor parameters."""
from makani.config import mconfig
from makani.control import system_types
import numpy as np


@mconfig.Config(deps={
    'gs': 'base_station.ground_station',
    'gs_model': 'base_station.gs_model',
})
def MakeParams(params):
  """Make wind sensor parameters."""
  if params['gs_model'] == system_types.kGroundStationModelTopHat:
    # Heading [rad] of the +u wind sensor axis relative to the +x axis
    # of the ground station coordinate system.
    heading = np.deg2rad(95.5)

    # Offset [m] of the wind sensor from the ground frame origin.
    pos = [0.202, -1.627, -3.808]

    on_perch = False

  elif params['gs_model'] == system_types.kGroundStationModelGSv1:
    # GSv1: These measurements are based off of the CAD drawings here:
    # go/makaniwiki/perch-geometry.
    #
    # The angle [rad] of the +u wind sensor axis relative to the +x axis
    # of the perch.
    heading = np.deg2rad(-30.0)
    pos = [-2.73, -2.35, -9.4]
    on_perch = True
  elif params['gs_model'] == system_types.kGroundStationModelGSv2:
    # 360 - 46.9 (ws to gps axis) - 169.84 (gps axis to the +x platform
    # coordinate system). See b/66695220.
    heading = np.deg2rad(143.26)
    pos = [2.32, 0.026, -14.9]
    on_perch = True
  else:
    assert False, 'Unsupported ground station model.'

  return {
      # Position [m] of wind sensor in the parent frame.
      'pos_parent': pos,

      # Orientation of the wind sensor relative to the parent frame.
      # Note that the +w axis of the wind sensor points upward.
      'dcm_parent2ws': {'d': [[np.cos(heading), np.sin(heading), 0.0],
                              [np.sin(heading), -np.cos(heading), 0.0],
                              [0.0, 0.0, -1.0]]},

      # Whether the wind sensor's parent frame is the perch frame, as opposed
      # to the ground frame.
      'on_perch': on_perch
  }

File: sim/models/sensors/wind_sensor.cc
// Copyright 2020 Makani Technologies LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "sim/models/sensors/wind_sensor.h"

#include <glog/logging.h>
#include <math.h>
#include <stdint.h>

#include "avionics/common/tether_message_types.h"
#include "common/c_math/vec3.h"
#include "common/macros.h"
#include "control/system_types.h"
#include "sim/faults/faults.h"
#include "sim/models/sensors/sensor.h"
#include "sim/models/signals/measurement.h"
#include "sim/physics/reference_frame.h"
#include "sim/sim_messages.h"
#include "sim/sim_telemetry.h"
#include "sim/sim_types.h"
#include "sim/state.h"

WindSensor::WindSensor(const WindSensorParams &wind_sensor_params,
                       const WindSensorSimParams &wind_sensor_sim_params,
                       FaultSchedule *faults)
    : Sensor("WindSensor"),
      wind_sensor_params_(wind_sensor_params),
      wind_sensor_sim_params_(wind_sensor_sim_params),
      wind_g_(new_input_value(), "wind_g"),
      ground_frame_(new_input_value(), "ground_frame"),
      parent_frame_(new_input_value(), "parent_frame"),
      wind_sensor_frame_(new_derived_value(), "wind_sensor_frame"),
      pos_g_(new_derived_value(), "pos_g"),
      sampled_wind_ws_(new_discrete_state(), "sampled_wind_ws", 0.0, kVec3Zero),
      measured_wind_ws_(full_name(), "measured_wind_ws",
                        wind_sensor_sim_params.sample_time, sampled_wind_ws_,
                        faults) {
  set_sub_models({&measured_wind_ws_});

  SetupDone();
}

void WindSensor::UpdateSensorOutputs(SimSensorMessage * /*message*/,
                                     TetherUpMessage *tether_up) const {
  TetherWind *wind = &tether_up->wind;
  wind->sequence =
      static_cast<uint16_t>((wind->sequence + 1) % TETHER_SEQUENCE_ROLLOVER);
  wind->no_update_count = 0;
  wind->status = kTetherWindStatusGood;
  wind->velocity[0] = static_cast<float>(measured_wind_ws().x);
  wind->velocity[1] = static_cast<float>(measured_wind_ws().y);
  wind->velocity[2] = static_cast<float>(measured_wind_ws().z);

  // TODO: Move this somewhere else.
  // TODO: Don't use hardcoded values.
  TetherWeather *weather = &tether_up->weather;
  weather->sequence =
      static_cast<uint16_t>((weather->sequence + 1) % TETHER_SEQUENCE_ROLLOVER);
  weather->no_update_count = 0;
  weather->flags = 0;
  weather->temperature = 25.0f;     // [C]
  weather->pressure_pa = 92060.0f;  // [Pa]
  weather->humidity = 7.0f;         // [%]
}

void WindSensor::Publish() const {
  sim_telem.wind_sensor.wind_g = wind_g();
  sim_telem.wind_sensor.measured_wind_ws = measured_wind_ws();
}

void WindSensor::UpdateDerivedStates() {
  wind_sensor_frame_.set_val(ReferenceFrame(parent_frame(),
                                            wind_sensor_params_.pos_parent,
                                            wind_sensor_params_.dcm_parent2ws));
  Vec3 pos_g__;
  wind_sensor_frame().TransformTo(ground_frame(), ReferenceFrame::kPosition,
                                  kVec3Zero, &pos_g__);
  pos_g_.set_val(pos_g__);
}

void WindSensor::AddInternalConnections(ConnectionStore *connections) {
  connections->Add(3, [this](double /*t*/) { UpdateDerivedStates(); });
}

void WindSensor::DiscreteStepHelper(double t) {
  // Transform the input wind velocity in the ground frame to
  // the wind sensor's local frame.  This accounts for the added
  // apparent wind from the moving perch.
  Vec3 wind_ws;
  wind_sensor_frame().TransformFrom(ground_frame(), ReferenceFrame::kVelocity,
                                    wind_g(), &wind_ws);

  sampled_wind_ws_.DiscreteUpdate(t, wind_ws);
}
